### ISSUE TITLE
Android 7: Trust anchor for certification path not found

### ISSUE THREAD

**Issue by seh-len** [owntracks/android#481]
Hi, 
since upgrading to Android 7.0 Nougat owntracks stopped working with my private mqtt. It shows the known error message. The cert definitely worked before and I even tried to add it to the cert store again. 

Cheers 
Simon

**Comment by binarybucks** [comment]
I'm working on Android 7 compatibility. 

There was a change regarding TLS contexts for user added CAs.

Meanwhile, could you add your CA directly in the app TLS settings? Maybe this can work around the problem.

**Comment by seh-len** [comment]
It doesn't work via app TLS settings, either.

**Comment by binarybucks** [comment]
That's strange because when you import the certificate in the app, it should work around the OS. 
Are you sure that the certificate chain is completely correct even though it worked before? 

I'm on the latest O beta too with a self signed root and intermediate cert and have not noticed any issues. 

However, I've added a new security config. I'll push it in the next beta for you to test. 
Maybe it fixes the issue.

**Comment by binarybucks** [comment]
Please report how build 12302 works for you.

**Comment by seh-len** [comment]
no, same error message with the beta version
mosquitto reports:
```
1495366493: OpenSSL Error: error:14094416:SSL routines:SSL3_READ_BYTES:sslv3 alert certificate unknown
1495366493: OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure
1495366493: Socket error on client <unknown>, disconnecting.
```

**Comment by jpmens** [comment]
@seh-len is there any chance that your TLS certificate has expired? (Sorry for asking this, but we've seen this before ... :-)

**Comment by jpmens** [comment]
I have `1.2.3 (12302)` running with a self-signed TLS certificate. The root CA is installed in Android's store, and OwnTracks on Android has no problem connecting to my private broker.

**Comment by binarybucks** [comment]
Please reopen if the issue persists

**Comment by seh-len** [comment]
My ca.crt is valid till 2025 and the same ca file is working fine with owntracks running on a different phone with android 6.0.1. But I'm not making use of a device cert file. Maybe this is the issue with android 7?

**Comment by binarybucks** [comment]
My self signed ca works flawlessly on my Android 7 device and in several emulators. 

Could you send the ca.crt in PEM format and the public address of your broker to support@owntracks.org?  I want to check if I can establish a TLS session with your broker.

**Comment by jpmens** [comment]
@seh-len thank you for CA cert and address. Using `openssl s_client`, `mosquitto_sub`, and our app, we get no error: the connection is correctly established, and we get past TLS negotiation. Obviously we don't get data because we're not authorized to, which is correct.

Have you changed anything in your broker setup? I ask because this doesn't correlate with your own experiences.

The only suspicious element in your CA cert is the `sha1WithRSAEncryption` signature algorithm, but we cannot find any reference to Android 7 no longer supporting SHA1; furthermore our tests with our App on Android 7 prove that that's ok.

**Comment by seh-len** [comment]
No, I didn't change anything, it just stopped working the moment I got the android 7 update on my Samsung S6.

**Comment by jpmens** [comment]
@seh-len I think we've done a lot to prove to you that we believe the app should be working for you. I don't see your effort in helping us track down your issue. Are you working at this or leaning back waiting for a "solution"?

**Comment by seh-len** [comment]
What do you mean with "working at this"? What are you expecting me to do? I'm using owntracks and it stopped working with android 7 and keeps working with another phone on android 6 with the same broker and the same ca file. So the only thing which have changed is the android version being used, and I reported this at this bug tracker. But feel free to close this issue because I just don't know how to help you in any way to track down this issue as there is only that single error message...

**Comment by jpmens** [comment]
Which Mosquitto version are you running on which OS/version?
Can you please show us (send to support@owntracks.org) without passwords, your `mosquitto.conf`?

Please also send us (without passwords) your owntracks app configuration (export).

We'll then see if we find anything

**Comment by binarybucks** [comment]
The weird thing is that I got the same trust anchor issue intially. 
Now @jpmens and I are able to connect without that error. Obviously we don't get past the broker authentication but the TLS session is established with your CA. 
This indicates that there is something going on with your broker. 

I suggest that you try to create a new SHA256 CA and broker certificate to test things out.

**Comment by seh-len** [comment]
After upgrading my CA and the server cert to sha256WithRSAEncryption it is working again.

**Comment by jpmens** [comment]
(For the record: I'm not convinced that's the issue.)

**Comment by otdftr** [comment]
having the same issue since I upgraded to Android 7 - all certs have Signature Algorithm: sha256WithRSAEncryption an 4096 bit keys
openssl verify succeeds 
tried the same client cert on an iPhone - owntracks 9.3.0_de - where it worked fine and connected to the mqtt service. same config, same destination, but on android 7 - connection files.

status says:
javax.net.SSL.HandshakeException: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found. 
I tried with and without CA-cert in the config (they are installed to the android tursted certs anyway), and I tried the client PKCS12 cert with and without certificate chain - no change.

server log says:
...: OpenSSL Error: error:14094416:SSL routines:SSL3_READ_BYTES:sslv3 alert certificate unknown
...: OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure
My server config demands tls_version tlsv1.2 - but even commenting that out or changing it to tlsv1 didn't change things: After a restart, all clients kept connecting successfully, ontracks @ Android kept failing.

**Comment by binarybucks** [comment]
Which app version are you using?
Is the ca cert imported in the app or the system ca store?

**Comment by binarybucks** [comment]
Might bei related to https://stackoverflow.com/questions/39133437/sslhandshakeexception-handshake-failed-on-android-n-7-0. 
Are you using any elliptic ciphers? That shouldn't lead to the trust anchor issue but I've seen stranger things.

**Comment by otdftr** [comment]
- sorry, forgot: owntracks 1.2.3. (12304) - most recent version from play store
- no, I do not use elliptic ciphers (at least not intentionally). checked the PKCS12 with openssl to make sure - for the private key, it shows "Shrouded Keybag: pbeWithSHA1And3-KeyTripleDES-CBC, Iteration 2048" 

to exclude any certificate issues, and to collect some more data, I extracted the certs / key from the pkcs12 and tried to connect to mosquitto with s_client:

```
$ openssl s_client -connect hydra:8883  -cert TEMPCERT.pem -CAfile TEMPCERT.pem -key CHECKEY.pem 
CONNECTED(00000003)SSL handshake has read 7737 bytes and written 7158 bytes
[...]
New, TLSv1/SSLv3, Cipher is AES256-GCM-SHA384
Server public key is 4096 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : AES256-GCM-SHA384
[...]
(connection established, waiting for input)
```
when trying to force SSL3, session setup fails:
```
$ openssl s_client -connect hydra:8883  -cert TEMPCERT.pem -CAfile TEMPCERT.pem -key CHECKEY.pem -ssl3
CONNECTED(00000003)
139781203179168:error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure:s3_pkt.c:1278:SSL alert number 40
139781203179168:error:1409E0E5:SSL routines:SSL3_WRITE_BYTES:ssl handshake failure:s3_pkt.c:599:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 0 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : SSLv3
    Cipher    : 0000
    Session-ID: Client connection from 192.168.66.226 failed: error:1408A10B:SSL routines:SSL3_GET_CLIENT_HELLO:wrong version number.

```
and mosquitto logs
````
Client connection from 192.168.66.226 failed: error:1408A10B:SSL routines:SSL3_GET_CLIENT_HELLO:wrong version number.
````

**Comment by otdftr** [comment]
__Suddenly, it works.__

Well. not suddenly: While testing & debugging another app, I found to "flaws" in the certificates that I did not see earlier - because the very same certificate combination worked well with other devices.
So while trying around with Android BKS key stores, I used the portecle java tool to create a key store, and that led me to the hidden issues:
- The server cert had a duplicate extension (X509v3 Basic Constraints: critical - CA:FALSE) which never caused any harm except with the java SSL SocketFactory. Don't now how it was even possible to build it - XCA is normally very strict about this.
- The client cert did not have X509v3 Subject Key Identifier / X509v3 Authority Key Identifier extensions - which IMHO should not matter, and BTW making a new cert with the KeyIDs did not fix the "peer did not return a certificate" error I had with the other app.

With the new certificates on Android and the mosquitto broker, owntracks works just fine. Right now, I am using a PKCS12 cert with embedded certificate chain and no CA cert in owntracks, but all my self-signed CA certificatess are in the android user cert store as well.

So I cant tell _what_ _exactly_ was the issue, but maybe this note helps in tracking down further / future issues... SSL can be a nightmare if it doesn't work right away...
